File: poc/ingredients.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>
#include <ctype.h>

typedef enum Unit_ {
	U_Unknown,
	U_Teaspoon,
	U_Tablespoon,
	U_CubicInch, //TODO
	U_Cup,
	U_Pint, //TODO
	U_Quart, //TODO
	U_Gallon, //TODO
	U_Ounce, //TODO
	U_FluidOunce, //TODO
	U_Pinch, //TODO
	U_Dash
} Unit;

/*
Key:
n - number (integer, decimal, fraction)
u - unit
i - ingredient
Text format:
n u i
nu i
*/

char *parseInteger(char *str, char *end, int *out) {
	int n = 0;
	while (str < end) {
		char c = *str;
		if (c < '0' || c > '9') break;
		
		n *= 10;
		n += c - '0';
		str++;
	}
	
	*out = n;
	return str;
}

/*
Parse a string as a quantity. May be an integer, decimal, or fraction.
*/
char *parseQuantity(char *str, char *end, float *out) {
	int part1 = 0, part2 = 0;
	str = parseInteger(str, end, &part1);
	
	//if next char is a point, parse this as a 2-part decimal number.
	if (str < end && *str == '.') {
		str++;
		char *part2Start = str;
		str = parseInteger(str, end, &part2);
		int nCharsPart2 = str - part2Start;
		
		*out = part1 + (part2 * pow(0.1, nCharsPart2));
	}
	
	//if next char is a /, parse this as a fraction.
	else if (str < end && *str == '/') {
		str++;
		str = parseInteger(str, end, &part2);
		*out = ((float) part1) / ((float) part2);
	}
	
	//if next char is a - or space, try to parse it as a mixed number
	else if (str < end && (*str == '-' || *str == ' ')) {
		char *backupStr = str;
		char sep = *str;
		int part3 = 0;
		str++;
		
		str = parseInteger(str, end, &part2);
		if (str < end && part2 > 0 && *str == '/') {
			str++;
			str = parseInteger(str, end, &part3);
			
			*out = part1 + (((float) part2) / ((float) part3));
		} else {
			//return integer (not valid mixed number)
			str = backupStr;
			*out = (float) part1;
		}
	}
	
	//else, return an integer
	else {
		*out = (float) part1;
	}
	
	return str;
}

int stringEqual(const char *s1, int nChars, const char *s2) {
	for (int i = 0; i < nChars; i++) {
		char c1 = s1[i];
		char c2 = s2[i];
		if (c2 == '\0') return 0;
		if (c1 != c2) return 0;
	}
	return 1;
}

char *parseUnit(char *str, char *end, Unit *outUnit) {
	//read chars until we hit a non-alphabetic character
	char *base = str;
	int nCharsUnit = 0;
	while (str < end && ((*str >= 'a' && *str <= 'z') || (*str >= 'A' && *str <= 'Z'))) str++;
	
	nCharsUnit = str - base;
	if (stringEqual(base, nCharsUnit, "tsp") || stringEqual(base, nCharsUnit, "teaspoon")
		|| stringEqual(base, nCharsUnit, "teaspoons")) {
		
		*outUnit = U_Teaspoon;
	} else if (stringEqual(base, nCharsUnit, "tbsp") || stringEqual(base, nCharsUnit, "tablespoon")
		|| stringEqual(base, nCharsUnit, "tablespoons")) {
		
		*outUnit = U_Tablespoon;
	} else if (stringEqual(base, nCharsUnit, "c") || stringEqual(base, nCharsUnit, "cup")
		|| stringEqual(base, nCharsUnit, "cups")) {
		
		*outUnit = U_Cup;
	} else {
		*outUnit = U_Unknown;
	}
	
	//if there's a dot after the unit, advance
	if (str < end && *str == '.') str++;
	
	return str;
}

char *parseRealQuantity(char *str, char *end, float *outAmount, Unit *outUnit) {
	char *start = str;
	float amount;
	Unit unit;
	
	//try parse quantity
	str = parseQuantity(str, end, &amount);
	
	//check that a quantity was parsed. 
	if (str > start) {
		//great. If next char is a space, advance.
		char *unitStart = str;
		while (str < end && *str == ' ') str++;
		
		//parse unit. agh
		str = parseUnit(str, end, &unit);
		if (str > unitStart) {
			if (str < end && *str == '.') str++;
			while (str < end && *str == ' ') str++;
			
			//if there's at least 3 characters to go, try skipping "of " if it exists.
			if (str + 3 <= end) {
				char c0 = toupper(str[0]);
				char c1 = toupper(str[1]);
				char c2 = toupper(str[2]);
				
				if (c0 == 'O' && c1 == 'F' && c2 == ' ') str += 3;
				while (str < end && *str == ' ') str++;
			}
			
			*outUnit = unit;
			*outAmount = amount;
			return str;
		}
	}
	
	//failure, backtrack
	str = start;
	
	*outUnit = U_Unknown;
	*outAmount = 0.0f;
	return str;
}

char *parseLine(char *str, char *end, float *outAmount, Unit *outUnit, char *outIngredient, int nMaxChars) {
	char *start = str;
	float amount = 0.0f;
	Unit unit = U_Unknown;
	
	//if line starts with *, -, or space, advance.
	while (str < end && (*str == '*' || *str == '-' || *str == ' ')) str++;
	
	str = parseRealQuantity(str, end, &amount, &unit);
	
	//check that a quantity was parsed. 
	if (str > start) {
		//leftover is ingredient
		while (nMaxChars > 0 && str < end && (*str != '\n' && *str != '\r')) {
			char c = *(str++);
			c = toupper(c);
			*(outIngredient++) = c;
			nMaxChars--;
		}
	} else {
		//
		//uh
	}
	
	*outUnit = unit;
	*outAmount = amount;
	return str;
}

int main(void) {
	char *str = "";
	float quant = 0;
	Unit unit;
	char ingredientBuffer[32] = {0};
	parseLine(str, str + strlen(str), &quant, &unit, ingredientBuffer, sizeof(ingredientBuffer) - 1);
	
	printf("%f %d of %s\n", quant, unit, ingredientBuffer);
	
	return 0;
}

File: poc/ingredients.py
#--------------------------------------------------------------------------------------------------
#
# Ingredient parser
#
# By Declan Moore, Evan Scherrer
# 
# Declan 03/07/2023: Ported the C algorithm to Python
#
#--------------------------------------------------------------------------------------------------

#of all the things Python has, it doesn't have strtol. Annoying!
def rp_parse_integer(string):
	val = 0
	for i in range(len(string)):
		chr = string[i]
		
		#if digit, continue
		int_value = ord(chr) - ord('0')
		if (int_value >= 0 and int_value <= 9):
			val *= 10
			val += int_value
		else:
			#not digit, break
			break
	
	#return a tuple: (value, remainder of string)
	return (val, string[i:])
	
"""
Parse a number in the beginning of a string. Can be a whole number, nonwhole number, or a mixed
number.

Parameters:
	string				String to parse

Returns:
	A tuple (number, string) of the parsed number and the remainder of the string unparsed.
"""
def rp_parse_quantity(string):
	part1 = 0
	part2 = 0
	(part1, string) = rp_parse_integer(string);
	
	#if next char is a point, parse this as a 2-part decimal number.
	if (len(string) > 0 and string[0] == '.'):
		string = string[1:]
		part2_start = string
		(part2, string) = rp_parse_integer(string);
		num_chars_part2 = len(part2_start) - len(string)
		
		val = part1 + (part2 * pow(0.1, num_chars_part2));
		return (val, string)
	
	#if next char is a /, parse this as a fraction.
	elif (len(string) > 0 and string[0] == '/'):
		string = string[1:]
		(part2, string) = rp_parse_integer(string);
		val = float(part1) / float(part2);
		return (val, string)
	
	#if next char is a - or space, try to parse it as a mixed number
	elif (len(string) > 0 and (string[0] == '-' or string[0] == ' ')):
		backup_str = string
		part3 = 0
		string = string[1:]
		
		(part2, string) = rp_parse_integer(string);
		if (len(string) > 0 and part2 > 0 and string[0] == '/'):
			string = string[1:]
			(part3, string) = rp_parse_integer(string)
			
			val = part1 + float(part2) / float(part3);
			return (val, string)
		else:
			#return integer (not valid mixed number)
			string = backup_str
			val = float(part1)
			return (val, string)
	
	#else, return an integer
	val = float(part1)
	return (val, string)
	
"""
Parses an ingredient string composed of a measurement and ingredient. The amount should appear
directly before the ingredient name in the string. It allows input of an integer, decimal number,
or a fixed number of the form "I N/D" or "I-N/D". Abbreviated unit names are converted into the
full singular form of the measurement name. On success, the function returns a 3-tuple of the 
amount, unit, and ingredient name. On failure, the function returns (0, None, None).

Parameters:
	string:				The string to parse

Returns:
	A tuple containing the parsed ingredient.
"""
def rp_parse_ingredient(string):
	#advance *, -, and spaces
	string = string.strip()
	while len(string) > 0 and (string[0] == '-' or string[0] == '*' or string[0] == ' '):
		string = string[1:]

	#get quantity
	(amount, string) = rp_parse_quantity(string.strip())
	
	#strip spaces
	string = string.lstrip()
	
	#tokenize
	tok = string.lower().split(" ")
	
	#if there's only one (or no) token, return error
	if (len(tok) <= 1):
		return (0, None, None)
	
	#dict of aliases and output names
	aliases = {
		#customary volumes
		"teaspoon": ["teaspoon", "teaspoons", "tsp", "tsps"],
		"tablespoon": ["tablespoon", "tablespoons", "tbsp", "tbsps"],
		"fluid ounce": ["floz", "fl", "fluid"], #if fl or fluid, consume another token
		"cup": ["cup", "cups", "c", "cs"],
		"quart": ["quart", "quarts", "qt", "qts"],
		"pint": ["pint", "pints", "pt", "pts"],
		"gallon": ["gallon", "gallons", "gal", "gals"],
		
		#customary weights
		"ounce": ["ounce", "ounces", "oz"],
		"pound": ["pound", "pounds", "lb", "lbs"],
		
		#metric volumes
		"liter": ["liter", "liters", "litre", "litres", "l"],
		"milliliter": ["milliliter", "milliliters", "millilitre", "millilitres", "ml",
			"cubic centimeter", "cubic centimeters", "cc", "ccs"],
		
		#metric weights
		"gram": ["gram", "grams", "g"],
		"kilogram": ["kilogram", "kilograms", "kg", "kgs", "kilo", "kilos"]
	}
	
	#equivalents for centimeter
	cm_aliases = ["cm", "cm.", "centimeter", "centimeters", "centimetre", "centimetres"]
	
	#equivalents for ounce
	oz_aliases = ["oz", "oz.", "ounce", "ounces"]
	
	#parse quantity. Mainly expand abbreviated unit names, and combine "fl oz" to one token
	unit_tok = tok[0]
	if unit_tok[-1] == ".":
		unit_tok = unit_tok[:-1]
	unit_name = None
	
	#scan aliases
	for name_entry in aliases.items():
		full_name = name_entry[0]
		names = name_entry[1]
		if unit_tok in names:
			unit_name = full_name
			break
	
	#if token isn't known, return an error
	if unit_name == None:
		return (0, None, None)
		
	#consume a token
	tok = tok[1:]
	if unit_tok == "fl" or unit_tok == "fluid":
		if len(tok) > 0 and (tok[0] in oz_aliases):
			tok = tok[1:]
		else:
			return (0, None, None)
	if unit_tok == "cubic":
		if len(tok) > 0 and (tok[0] in cm_aliases):
			tok = tok[1:]
		else:
			return (0, None, None)
	
	#if next token is "of", skip
	if len(tok) > 0 and tok[0] == "of":
		tok = tok[1:]
	
	return (amount, unit_name, " ".join(tok))
	
